### PR TITLE
use local copy of less.min.js

### PR DESCRIPTION
**Pull request by mib1185**
### Description

This PR adds a local copy of https://cdnjs.cloudflare.com/ajax/libs/less.js/2.7.2/less.min.js and set it to be used localy, instead of always downloading a static file from the web.
This Fixes: #5277

[less.js is also under Apache-2.0 License](https://github.com/less/less.js/blob/master/LICENSE), those no issue expected when distribute the file directly.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

prior this change:
```
$ grep "less.min.js" ui/dist/index.html 
          <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/less.js/2.7.2/less.min.js"></script>
$ ls ui/dist/js/less.min.js 
ls: cannot access 'ui/dist/js/less.min.js': No such file or directory
```

with this change:
```
$ grep "less.min.js" ui/dist/index.html 
          <script type="text/javascript" src="js/less.min.js"></script>
$ ls ui/dist/js/less.min.js 
ui/dist/js/less.min.js
```

### CODE REVIEW

**Comment by mib1185** [comment]
@blueorangutan ui

**Comment by rhtyd** [comment]
cc @utchoang @davidjumani @Pearl1594 @shwstppr
Do we really need this, or can we get less to bundle on build time?

**Comment by mib1185** [comment]
@rhtyd for us it is a major issue in UX, since our management environment (_based on Citrix XenApp_) is not allowed to connect to the internet, those on each page reload/open new tab the browser hangs until a timeout occurred while trying to load https://cdnjs.cloudflare.com/ajax/libs/less.js/2.7.2/less.min.js (_Firefox ~60s, Chrome ~40s_)

**Review by nvazquez — COMMENTED** [review]
Looks good, however is not clear to me which version to use in case this file needs to be updated in future

**Review by utchoang — APPROVED** [review]
LGTM!

File: ui/theme.config.js [hunk 1]
```diff
@@ -18,7 +18,7 @@
 const path = require('path')
 const AntDesignThemePlugin = require('antd-theme-webpack-plugin')
 
-function resolve (dir) {
+function resolve(dir) {
```
Comment (shwstppr) [inline ui/theme.config.js:21]: Not needed

**Comment by shwstppr** [comment]
> cc @utchoang @davidjumani @Pearl1594 @shwstppr
> Do we really need this, or can we get less to bundle on build time? And does this apply for 4.15.2?

@rhtyd could be done using  pre/post build scripts but I'm not sure if that would add any value,

```
diff --git a/ui/postbuild.sh b/ui/postbuild.sh
index ab7734f383..72c25c8a4e 100755
--- a/ui/postbuild.sh
+++ b/ui/postbuild.sh
@@ -19,6 +19,16 @@
 DIR=$(dirname $0)
 configFile="$DIR/public/config.json"
 tmpFile="$DIR/public/config.json.tmp"
+jsDir="$DIR/public/js"
+lessJsFile="$jsDir/less.min.js"
+echo "Post-build: Removing ${lessJsFile}"
+if [ -f $lessJsFile ]; then
+   rm $lessJsFile
+fi
+if ! rmdir $jsDir 2> /dev/null; then
+    echo "$jsDir contains the following files:"
+    ls $jsDir
+fi
 echo "Post-build: removing all docHelp suffixes in ${configFile}"
 
 node > ${tmpFile} <<EOF
diff --git a/ui/prebuild.sh b/ui/prebuild.sh
index 771daa84b2..00cd611906 100755
--- a/ui/prebuild.sh
+++ b/ui/prebuild.sh
@@ -19,6 +19,15 @@
 DIR=$(dirname $0)
 configFile="$DIR/public/config.json"
 tmpFile="$DIR/public/config.json.tmp"
+jsDir="$DIR/public/js"
+lessJsFile="$jsDir/less.min.js"
+lessJsUrl="https://cdnjs.cloudflare.com/ajax/libs/less.js/2.7.2/less.min.js"
+echo "Pre-build: Creating ${lessJsFile}"
+if [ ! -d $jsDir ]; then
+  mkdir $jsDir
+fi
+wget -O $lessJsFile $lessJsUrl
+
 echo "Pre-build: list all docHelp suffixes in ${configFile}"
 for m in $(grep "docHelp: '" -R ./src | sed "s/^.*: '//g" | sed "s/',//g" | sort | uniq); do
     docHelpMappings+="${m},"
diff --git a/ui/theme.config.js b/ui/theme.config.js
index 8a41c49b5c..6bdd891bc7 100644
--- a/ui/theme.config.js
+++ b/ui/theme.config.js
@@ -50,6 +50,7 @@ const options = {
   ],
   indexFileName: 'index.html',
   publicPath: '.',
+  lessUrl: 'js/less.min.js',
   generateOnce: false
 }
```

**Review by davidjumani — APPROVED** [review]

**Comment by shwstppr** [comment]
Screenshot from 2021-09-07 18-02-55
UI loads local copy

But not sure why I'm seeing a difference in the added file and the file got from https://cdnjs.cloudflare.com/ajax/libs/less.js/2.7.2/less.min.js or https://raw.githubusercontent.com/less/less.js/v2.7.2/dist/less.min.js

**Comment by rhtyd** [comment]
Merging this manually on 4.15 branch and then on main branch.

**Comment by rhtyd** [comment]
@shwstppr thanks I've used src from the cloudflare URL to be safe. the one from Cloudflare is minimised.

**Comment by rhtyd** [comment]
Commit ref - https://github.com/apache/cloudstack/commit/c0ecc34f4c8d8cfb4ae0f619a2b21330a1c145ca

**Comment by mib1185** [comment]
> But not sure why I'm seeing a difference in the added file and the file got from https://cdnjs.cloudflare.com/ajax/libs/less.js/2.7.2/less.min.js or https://raw.githubusercontent.com/less/less.js/v2.7.2/dist/less.min.js

it figured out, that this was because of the auto-formatting from VS Code

**Pull request closed without merging**